### ISSUE TITLE
Problem importing previous training data when using gaussian process

### ISSUE THREAD

**Issue by sebbleroni** [michaelhush/M-LOOP#82]
**Describe the bug**
I am trying to use data from previous experiments as input for my optimizer. Unfortunately this does not work if I use a Gaussian Process controller (yet, for a Differential Evolution controller everything seems to work). Is this a known issue and is there a workaround?

**To Reproduce**
1. create a simple controller and optimize (for example the optimizer from the tutorial):

```python
#Imports for M-LOOP
import mloop.interfaces as mli
import mloop.controllers as mlc
import mloop.visualizations as mlv
import mloop.utilities as mlu

class CustomInterface(mli.Interface):
    
    def __init__(self):
        super(CustomInterface,self).__init__()
        self.minimum_params = np.array([0,0.1,-0.1])

    def get_next_cost_dict(self,params_dict):
        
        params = params_dict['params']
        
        cost = -np.sum(np.sinc(params - self.minimum_params))
        uncer = 0
        bad = False
        
        cost_dict = {'cost':cost, 'uncer':uncer, 'bad':bad}
        return cost_dict
    
def main():
    
    input_dict = mlu.get_dict_from_file('config.txt', 'txt')
    interface = CustomInterface()
    controller = mlc.create_controller(interface, **input_dict)
    controller.optimize() 

if __name__ == '__main__':
    main()
```

with the configuration file
```python
max_num_runs = 15
target_cost = -2.99
num_params = 3
min_boundary = [-2,-2,-2]
max_boundary = [2,2,2]
controller_type = 'differential_evolution'
param_names=None

training_type = 'differential_evolution'   
gp_training_filename = "training_data.txt"
gp_training_file_type = 'txt'        
```

where I either use `controller_type = 'differential_evolution'` or `controller_type = 'gaussian_process'`.  `training_data.txt` simply is a learner archive from a previous run.

**Expected behavior**

- previous run  `controller_type = 'differential_evolution'` and current run  `controller_type = 'differential_evolution'`: works
- previous run  `controller_type = 'gaussian_process'` and current run  `controller_type = 'differential_evolution'`: works
- previous run  `controller_type = 'gaussian_process'` and current run  `controller_type = 'gaussian_process'`: `TypeError: __init__() got multiple values for keyword argument 'param_names'`
- previous run  `controller_type = 'differential_evolution'` and current run  `controller_type = 'gaussian_process'`: `TypeError: __init__() got multiple values for keyword argument 'param_names'`

**Screenshots**
![image](https://user-images.githubusercontent.com/70020179/102093391-65576000-3e21-11eb-8d49-85c952d85a7a.png)

**Desktop (please complete the following information):**

- OS: Ubuntu 16.04
- M-LOOP version 3.1.1
- Python version 3.5.2

**Comment by charmasaur** [comment]
@zakv FYI

Issue seems to be that [here](https://github.com/michaelhush/M-LOOP/blob/master/mloop/learners.py#L1020) we're both passing `param_names` explicitly (with the values we extracted from the training dict [here](https://github.com/michaelhush/M-LOOP/blob/master/mloop/learners.py#L967) and passing it via the `kwargs`). I guess there are a few ways we could deal with it, one of which would be:
 - explicitly list `param_names` in the `GaussianProcessLearner` constructor (which gets it out of `kwargs),
 - only set `self.param_names` from the training dict if the `param_names` are `None` (which seems to me the right direction for the override to go),
 - and pass `param_names` explicitly when calling the superclass constructor in the case that no training dict is provided.

Any better ideas, Zak?

---

@sebbleroni sorry for the trouble. We'll work on getting this fixed properly, but in the meantime a few workarounds come to mind, any of which should do the trick:
1. Locally implement the changes described above (I haven't actually tested them, but they should work I think); you could either do this by installing from source, or just editing `.../site-packages/mloop/learners.py` directly.
2. Delete the `param_names` from `training_data.txt`.
3. Locally implement a hackier fix, for example delete the line [here](https://github.com/michaelhush/M-LOOP/blob/3.1.1/mloop/learners.py#L1000) (again, either after installing from source or just by editing the file in `site-packages`).

**Comment by zakv** [comment]
Thanks @sebbleroni for the detailed report, and thanks @charmasaur for the looking into the issue!

I think your suggestions are the correct way to go Harry, and they should be applied to `NeuralNetLearner.__init__()` as well. I'll get to work implementing them.

As for the workarounds, I think Harry's suggestion #2 is probably the way to go for now. You may want to make a backup copy of `training_data.txt` before editing it though.

**Comment by zakv** [comment]
Actually this issue happens not only for `param_names` but for all of the other arguments [explicitly passed here](https://github.com/michaelhush/M-LOOP/blob/bc172f251ef1fb71ca6c2df1f9d82f20f0eb7191/mloop/learners.py#L1016-L1019) as well. I think the simplest way to handle this might just be to modify the `kwargs` dict as necessary, adding key/value pairs for parameters not already in it, and raising a `ValueError` if the keyword argument values provided don't match those in the training archive. Then all of the values will be passed to the parent class via `__init__(**kwargs)` without any other arguments explicitly being specified.

Also, using the learner archive from a differential evolution learner optimization as the training archive for a gaussian process learner optimization is a little trickier. The `GaussianProcessLearner.__init__()` is written assuming that the archive will have all of the gaussian-learner-related attributes, such as `'cost_has_noise'`, so it errors out when using a differential evolution learner archive which doesn't have that entry. Maybe it would be enough to just set those properties only if they exist in the archive and do nothing otherwise; I haven't thought it through yet.

It's maybe worth mentioning that the differential evolution learner works for the "current run" in @sebbleroni's test because it doesn't use the training file. That learner doesn't support a training file option, and in general M-LOOP just silently ignores any extra options that it doesn't know how to interpret.

**Comment by sebbleroni** [comment]
> Thanks @sebbleroni for the detailed report, and thanks @charmasaur for the looking into the issue!
> 
> I think your suggestions are the correct way to go Harry, and they should be applied to `NeuralNetLearner.__init__()` as well. I'll get to work implementing them.
> 
> As for the workarounds, I think Harry's suggestion # 2 is probably the way to go for now. You may want to make a backup copy of `training_data.txt` before editing it though.
> 
> edit: removed accidental link to unrelated issue number 2

that was also my go to option, but then I realized the same thing a Zak, namely that this happens for many parameters. I am generally reluctant to modify my local version of M-LOOP. But maybe it's necessary in this case...

Thank you all for being so responsive and helpful! (and sorry already for the next issue I raised)

**Comment by zakv** [comment]
@sebbleroni This issue was automatically closed when #84 was merged. This was fixed in my tests but feel free to re-open this issue if you still have problems!

Using learner archives generated by one learner as the training archive for a new optimization with a different type of learner isn't implemented though. If you need that feature you're welcome to take a swing at implementing it! I can try to provide some guidance if you do want to give it a go.

**Comment by zakv** [comment]
Actually #90 should now make it possible to use archives from a learner of any type as the training archive for a neural net or Gaussian process optimization. Note that it won't work on old archives from other learners though; they must have been generated with the changes in that PR.
